File: pset1/credit/credit_v1/credit.c
#include <stdio.h>
#include <cs50.h>

int main(void)
{
    // declare credit card number
    long card_number;

    int counter = 0; // counter of digits in card_number
    long card_copy; // for calculations in loops
    int firsts; // for calculate sum of every first number
    int seconds; // for calculate sum of every other number
    int first_digit; // for store firsts digits
    int second_digit; // for store seconds digits
    int sum1 = 0; // sum of firsts digits
    int sum2 = 0; // sum of every other number
    int sum3 = 0; // final sum

    // ask user for card number (but no validation here because of CS50 checklist requirements - but that is stupid as for me)
    card_number = get_long("write card number: ");

    // count card number of digits
    card_copy = card_number;
    while (card_copy > 0)
    {
        card_copy /= 10;
        counter++;
    }

    // restore card number to card copy after previous loop
    card_copy = card_number;

    // calculate check sum
    while (card_copy > 0)
    {
        {
            firsts = card_copy % 10;
            card_copy /= 10;
        }

        // here we get sum of all firsts digits from end
        sum1 = sum1 + firsts;

        {
            seconds = card_copy % 10;
            card_copy /= 10;
        }

        // here we get sum of all seconds digits from end
        seconds *= 2;
        if (seconds > 9)
        {
            seconds = 1 + (seconds % 10);
        }

        sum2 = sum2 + seconds;
    }

    // here we get final sum
    sum3 = sum1 + sum2;

    // get last digit in sum3
    sum3 %= 10;

    // restore card number to card copy after previous loop
    card_copy = card_number;

    // define first digit of card number
    first_digit = 0;
    while (card_copy > 0)
    {
        first_digit = card_copy % 10;
        card_copy /= 10;
    }

    // restore card number to card copy after previous loop
    card_copy = card_number;

    // define second digit of card number
    second_digit = 0;
    while (card_copy > 9)
    {
        second_digit = card_copy % 10;
        card_copy /= 10;
    }

    // validate first digit 4 - VISA

    if (((counter == 13) || (counter == 16)) && (sum3 == 0) && (first_digit == 4))
    {
        printf("VISA\n");
    }
    // validate first digit 5, second digit 1,2,3,4,5 - mastercard
    else if ((counter == 16) && (sum3 == 0) && (first_digit == 5 && (second_digit >= 1 & second_digit <= 5)))
    {
        printf("MASTERCARD\n");
    }
    // validate first digit 3, second digit 4, 7 - AMEX
    else if ((counter == 15) && (sum3 == 0) && (first_digit == 3 && (second_digit == 4 || second_digit == 7)))
    {
        printf("AMEX\n");
    }
    else
    {
        printf("INVALID\n");
    }
}

File: pset6/credit/credit.py
# OK. let's make our credit problem in python
# this is detailed explanation for beginners in programing like me:)

# first of all let's open our main function
def main():
    cardNumber = getCardNumber()  # let's get valid card number from user
    checkSum = calculateCheckSum(cardNumber)  # and calculate cards number check sum
    checkCardType(cardNumber, checkSum)  # check card type and print it


def getCardNumber():  # let's ask user to type in a valid card number
    while True:
        cardNumber = input("Valid card number: ")
        if cardNumber.isnumeric():  # using `.isnumeric()` method to check if all ipnut contains numbers
            break  # if valid -> stop asking
    return cardNumber  # and return its number


def calculateCheckSum(cardNumber):  # let's try another way compare to pset1
    evenNumberSum = 0  # we're gonna count sum of all even numbers
    oddNumberSum = 0  # and sum of all odd numbers
    cardNumber = reversed([int(currentNumber) for currentNumber in cardNumber])  # create `reversed()` copy of card number

    # enumerate() - method adds counter to an iterable and returns it (the enumerate object)
    for i, currentNumber in enumerate(cardNumber):
        # how to check if current number is odd?
        # add to it 1 and then divide by 2 (use modulo %2) and check if remainder is 0 or 1
        if (i + 1) % 2 == 0:
            oddNumber = currentNumber * 2  # Luhn's Algorithm 1. Multiply every other digit by 2
            # if its value greater than 9 (hsa two digits) -> add those digits together
            if oddNumber > 9:
                # for exmple:
                # number 12 contains two digits - 1 and 2
                # how to get 1 from 12?
                # 12 / 10 = 1.2
                # and use method -> int(1.2) = 1 - that give us int from float
                # and how to get 2 from 12?
                # use modulo 12 % 10 that give us remainder 2
                oddNumberSum += int(oddNumber / 10) + oddNumber % 10
            else:  # if our value is lexx than 9 -> just add it to odd numbers sum
                oddNumberSum += oddNumber
        else:  # if currentNumber is even
            evenNumberSum += currentNumber  # just add it to even number sum
    checkSum = evenNumberSum + oddNumberSum  # Luhn's Algorithm 2. Add the sum to the sum of the digits that weren’t multiplied by 2

    # and return card's check sum
    return checkSum


def checkCardType(cardNumber, checkSum):  # now let's make a function that check users card type and validity
    firstDigit = int(cardNumber[0:2])  # we have to check first digits of card number
    cardNumberLength = len(cardNumber)  # and check how many digits in card number
    isLastDigitZero = checkSum % 10  # and check last digit from check sum be `0`

    # Luhn's Algorithm 3. If the total’s last digit is 0 (or, put more formally, if the total modulo 10 is congruent to 0), the number is valid!
    if isLastDigitZero == 0:  # if last digit is zero then card number is valid and can be checked for type
        if firstDigit in [34, 37] and cardNumberLength == 15:  # check for AMEX (first digits 34 or 37 and card length is 15)
            print("AMEX")
        elif (int(cardNumber[0]) == 4) and cardNumberLength in [13, 16]:  # check for VISA (first digit is 4 and card length is 13 or 16)
            print("VISA")
        elif (firstDigit in range(51, 56)) and cardNumberLength == 16:  # check for MASTERCARD (first digits is from 51 to 56 and card length is 16)
            print("MASTERCARD")
    else:  # if not valid card length
        print("INVALID")


main()  # and close our main function